oxd-server/src/main/java/org/gluu/oxd/server/op/Validator.java

```diff
@@ -142,14 +143,22 @@ public void validateAuthorizationCode(String code) {
     public static AbstractJwsSigner createJwsSigner(Jwt idToken, OpenIdConfigurationResponse discoveryResponse, PublicOpKeyService keyService, OpClientFactory opClientFactory, Rp rp) {
         final String algorithm = idToken.getHeader().getClaimAsString(JwtHeaderName.ALGORITHM);
         final SignatureAlgorithm signatureAlgorithm = SignatureAlgorithm.fromString(algorithm);
-        final String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
         final String jwkUrl = discoveryResponse.getJwksUri();
+        String kid = idToken.getHeader().getClaimAsString(JwtHeaderName.KEY_ID);
 
         if (signatureAlgorithm == null)
             throw new HttpException(ErrorResponseCode.INVALID_ALGORITHM);
 
+        if (Strings.isNullOrEmpty(kid) && (signatureAlgorithm.getFamily() == AlgorithmFamily.RSA || signatureAlgorithm.getFamily() == AlgorithmFamily.EC)) {
```

Comment:
If `kid` is missing lets reject it. Conformance test says that it's up to RP. We should require kid be specified.

Reply:
as discussed on chat, we need `getKeyId` method to pass the conformance test. The test should pass if `kid` not in `id_token` and if `key` present in keystore.